test/services/GasEstimator.test.ts

```diff
@@ -0,0 +1,233 @@
+import { expect } from 'chai';
+import sinon, { StubbedInstance, stubConstructor } from 'ts-sinon';
+import { BigNumber, ethers } from 'ethers';
+
+import { GasEstimator } from '../../src/services/GasEstimator';
+import { GasEstimation } from '../../src/types/GasEstimation';
+import { BlockWithTransactions, FeeData, TransactionResponse } from '@ethersproject/abstract-provider';
+
+describe('GasEstimator', () => {
+  describe('.apply', () => {
+    const fakeProvider: StubbedInstance<ethers.providers.BaseProvider> = stubConstructor(ethers.providers.BaseProvider);
+
+    beforeEach(() => {
+      fakeProvider.getGasPrice.resolves();
+      fakeProvider.getBlockWithTransactions.resolves();
+      fakeProvider.getFeeData.resolves();
+    });
+
+    afterEach(() => {
+      sinon.restore();
+    });
+
+    before(() => {
+      GasEstimator.apply(fakeProvider, 1e6, 1e9).then();
+    });
+
+    it('calls getGasPrice from Provider', () => {
+      expect(fakeProvider.getGasPrice.called).to.be.true;
+    });
+
+    it('calls getBlockWithTransaction from Provider', () => {
+      expect(fakeProvider.getBlockWithTransactions.called).to.be.true;
+    });
+
+    it('calls getFeeData from Provider', () => {
+      expect(fakeProvider.getFeeData.called).to.be.true;
+    });
+
+    describe('when provider fails to retrieve data', () => {
+      beforeEach(() => {
+        fakeProvider.getGasPrice.rejects();
+        fakeProvider.getBlockWithTransactions.rejects();
+        fakeProvider.getFeeData.rejects();
+      });
+
+      afterEach(() => {
+        sinon.restore();
+      });
+
+      it('throws an error', () => {
```

Comment:
I think it should be async function

Reply:
Chai tests async functions differently. We use `.to.eventually` to expect asynchronous to be resolved.

Reply:
In this case, rejected*.
I was taking a look at some forums and this was the better way I found of doing it

Reply:
Looks good!